hyperspy/io_plugins/zspy.py

```diff
@@ -0,0 +1,237 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+#  HyperSpy is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with  HyperSpy.  If not, see <http://www.gnu.org/licenses/>.
+
+import warnings
+import logging
+from functools import partial
+from collections import MutableMapping
+
+import zarr
+from zarr import Array, Group
+import numpy as np
+import dask.array as da
+from hyperspy.io_plugins.hspy import version
+import numcodecs
+
+
+from hyperspy.io_plugins.hierarchical import HierarchicalWriter, HierarchicalReader, _overwrite_dataset
+
+_logger = logging.getLogger(__name__)
+
+
+# Plugin characteristics
+# ----------------------
+format_name = 'ZSpy'
+description = \
+    'A default file format for HyperSpy based on the zarr standard'
+full_support = False
+# Recognised file extension
+file_extensions = ['zspy']
+default_extension = 0
+# Writing capabilities
+non_uniform_axis = True
+writes = True
+
+# -----------------------
+# File format description
+# -----------------------
+# The root must contain a group called Experiments
+# The experiments group can contain any number of subgroups
+# Each subgroup is an experiment or signal
+# Each subgroup must contain at least one dataset called data
+# The data is an array of arbitrary dimension
+# In addition a number equal to the number of dimensions of the data
+# dataset + 1 of empty groups called coordinates followed by a number
+# must exists with the following attributes:
+#    'name'
+#    'offset'
+#    'scale'
+#    'units'
+#    'size'
+#    'index_in_array'
+# The experiment group contains a number of attributes that will be
+# directly assigned as class attributes of the Signal instance. In
+# addition the experiment groups may contain 'original_metadata' and
+# 'metadata'subgroup that will be
+# assigned to the same name attributes of the Signal instance as a
+# Dictionary Browsers
+# The Experiments group can contain attributes that may be common to all
+# the experiments and that will be accessible as attributes of the
+# Experiments instance
+
+
+def get_object_dset(group, data, key, chunks, **kwds):
+    """Overrides the hyperspy get object dset function for using zarr as the backend
+    """
+    these_kwds = kwds.copy()
+    these_kwds.update(dict(dtype=object,
+                           exact=True,
+                           chunks=chunks))
+    dset = group.require_dataset(key,
+                                 data.shape,
+                                 object_codec=numcodecs.VLenArray(int),
+                                 **these_kwds)
+    return dset
+
+
+def _get_signal_chunks(shape, dtype, signal_axes=None):
```

Comment:
Why not reusing the the `hyperspy.io_plugins.hspy.get_signal_chunks` and move it to `hyperspy.io_plugins.hierarchical?

Reply:
This is based on the how to most efficiently save using default signal chunking.  HDF5 tries to make chunks that are around 1 mb where zarr says that an ideal chuck size is 100 mb. The

Reply:
Would it be possible to make the target size a parameter and use only one function?